### PR TITLE
feat: New GCP Pub/Sub source

### LINKED ISSUE #753: New `gcp_pubsub` source
Hello,
Vector supports Kafka and aws Kinesis (where vector agents publish events to)
It will be great to add support for Google pub/sub
Why?

1. GCP and pub/sub specifically is gaining more and more popularity , especialy in the data processing world
2. Integrate pub/sub with Vector consumer
3. Google BiqQuery is extremely popular so streaming events to it will be very beneficial for the Vector adoption

Thanks
Vladi Feigin

### PR DESCRIPTION
**Pull request by bruceg**
Closes #753

~This isn't _quite_ ready for merging yet, as acknowledgement handling is not yet tested, but the rest of the structure is complete and ready for initial reviews.~ This should be ready for final review.

### CODE REVIEW

File: LICENSE-Apache-2.0.txt
Comment (binarylogic) [inline LICENSE-Apache-2.0.txt:1]: What's this about?

**Review by jdrouet — APPROVED** [review]
Besides the failing tests, the code looks good to me 👍

File: LICENSE-Apache-2.0.txt
Comment (bruceg) [reply LICENSE-Apache-2.0.txt:1]: The GCP protobuf definitions are copied straight out of Google's public repository, and they carry the APL on them, which requires the presence of that license file.

Comment (spencergilbert) [reply LICENSE-Apache-2.0.txt:1]: I'd expect we'd need to make it clear it just applies to the protobuf files, no? (of course I know nothing about license law)

Comment (binarylogic) [reply LICENSE-Apache-2.0.txt:1]: Yes, we should place this in the notices file or somewhere that is not the root of the project. Otherwise it'll get confused with out own license.

Comment (bruceg) [reply LICENSE-Apache-2.0.txt:1]: Would it be worthwhile to try to rewrite the proto definitions from the online docs instead of using this (minimized) copy from the protobuf repository? Another option might be to pull in the API repository as a git submodule, although this would be the precedent for that.

**Comment by rrs45** [comment]
Thanks guys! eagerly waiting for this to land. Any estimate when this will be available in release?

**Comment by jszwedko** [comment]
Hi @rrs45 !

It'll likely be in 0.22.0 in ~4 weeks. It'll be in a nightly release earlier though.

**Comment by bruceg** [comment]
I have worked out the shutdown issue, so this should be ready for final review. I have one small test worry that I am chasing down, but I don't expect it to majorly change the code.

File: src/sources/gcp_pubsub.rs
Comment (spencergilbert) [inline src/sources/gcp_pubsub.rs:419]: These tests could probably use more descriptive names

**Review by tobz — APPROVED** [review]
LGTM overall, modulo Spencer's comments about wanting slightly more descriptive unit test names, which I agree with.

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**
